### DISCUSSION TITLE
Directory label for 1,2, 3, 4 tabs?

### DISCUSSION THREAD

**Discussion by Praful** [jarun/nnn#1914] [Ideas]
The addition of tabs to nnn is a great feature to an already excellent minimalist file manager. I'm wary of feature creep that turns good software bad!

Whilst it's good to have the four default tabs with the numbers serving as shortcuts, I often forget which directory is which. Maybe it's my age! I'm constantly cycling through the tabs to find the one I want. This concealed information (ie the name of the directory the tab represents) could be made explicit by adding the directory to the tab name, eg, "1 - ~/Documents", "2 - /mnt/backup", etc.

Although less vital for me, could pressing 5-9 add new tabs?

Thank you for providing the program.

**Comment by jarun** [comment]
Looks like you are just getting started with `nnn`. Please read the wiki.

You can compile-in up to 8 contexts: https://github.com/jarun/nnn/wiki/Developer-guides#compile-in-8-contexts
It seems by default X11 integration is available: https://github.com/jarun/nnn/wiki/Developer-guides#compile-out-x11-integration
If it is available, use the program option `-x` to advertise the current directory to xterm: https://github.com/jarun/nnn/wiki/Usage#program-options

**Comment by Praful** [comment]
@jarun thanks for the links and explanation. I had used `-x`, which sets the tab title in `kitty`. Note this doesn't work for me (in `kitty` or `GNOME Terminal`) if `-S` is also specified, ie `nnn -x` sets title on folder change,  `nnn -x -S` doesn't set at all. I've tried with the pre-compiled v4.3 and locally compiled 4.9.

I'd like the name of the folders within `nnn` itself. For example, currently, when I start `nnn`, I see on the top:

```
1 2 3 4 ~/
```
where `~/` is the folder of the current context, say `1`. If I change to context `2`, the folder to the right of `4` changes to the new folder. In other words, `nnn` displays the folder only of the current context. I'd like to see the folders each context represents without having to switch to them first. So, something like:
```
1 = ~/, 2 = ~/documents, 3, 4 
```
Without formatting the above looks confusing but I hope you get the idea: context `1` is `~/`, context `2` is `~/documents` and contexts 3 and 4 have not been activated. This is, to me, a more transparent way of representing the contexts and saves me having to cycle through the contexts (which is admittedly fast) to find the folder I want.

**Reply by jarun** [reply]
That's a lot of change and very subjective.
Right now you can press `?` and see the dirs for each context in the LOCATIONS section.

**Reply by Praful** [reply]
Please correct my understanding. To me, the term "contexts" seems to be, essentially, what other apps call "tabs". There may be some subtle differences but calling them tabs fits my mental model of what `nnn` is trying to do with the feature.

If contexts are essentially tabs, then the "1 2 3 4" that appears at the top of `nnn` tells you that you can have up to 4 tabs, with the first one "activated" by default and others activated by pressing the `tab` or a number key.

Labelling the pre-defined tabs "1 2 3 4" hints at telling the user that you can press `1` to open tab 1, `2` to open tab 2, etc. That's a good shortcut and pressing the `tab` key also works. "1 2 3 4" is also short.

However, every other app that I've used which has tabs, gives the tab a name that reflects its contents. In various file managers (eg File Explorer in Windows and nemo or thunar on some Linux systems) the tab name is the name of the current directory. In editors, the tab name is the name of the file being edited. It would be strange to have editor tabs, say, labelled "1, 2, 3, ..." and for the user to have to look up what these tabs represent. And yet that is what `nnn` does. This seems to me an unnecessary indirection. But, again, correct me if I'm undestanding this wrong.

Of course, I could press `?`, as you suggest, make a note of which number I want to go to then navigate to that tab. It would be quicker just to keep pressing the `tab` key until I get to my desired directory. This all seems unnecessarily convoluted when you can simply tell the user what "1 2 3 4" represents!

To summarise, my suggestion is that instead of labelling the tabs ("contexts") "1 2 3 4", you label them with the names of directories they represent. This may, as you said, be "subjective" but it's what every other app with tabs that I use does.

If "contexts" are essentially tabs, then it would be clearer to new users if the documentation called them "tabs" rather than "contexts", which to me seems too vague and therefore less meaningful.

**Reply by N-R-K** [reply]
Yes, contexts are essentially tabs.

> To summarise, my suggestion is that instead of labelling the tabs ("contexts") "1 2 3 4", you label them with the names of directories they represent.

I tried something like this out and it just looks noisy.

```diff
diff --git a/src/nnn.c b/src/nnn.c
index ccbebe74..66e51bda 100644
--- a/src/nnn.c
+++ b/src/nnn.c
@@ -6644,10 +6644,17 @@ static void redraw(char *path)
 			addch((i + '1') | (COLOR_PAIR(i + 1) | A_BOLD
 				/* active: underline, current: reverse */
 				| ((cfg.curctx != i) ? A_UNDERLINE : A_REVERSE)));
+		if (g_ctx[i].c_cfg.ctxactive) {
+			attron (COLOR_PAIR(i + 1) | ((cfg.curctx != i) ? A_UNDERLINE : A_REVERSE));
+			addstr(" = ");
+			addstr(g_ctx[i].c_path);
+			attroff(COLOR_PAIR(i + 1) | ((cfg.curctx != i) ? A_UNDERLINE : A_REVERSE));
+		}
 
 		addch(' ');
 	}
 
+#if 0
 	attron(A_UNDERLINE | COLOR_PAIR(cfg.curctx + 1));
 
 	/* Print path */
@@ -6691,6 +6698,7 @@ static void redraw(char *path)
 		reset_tilde_in_path(path);
 
 	attroff(A_UNDERLINE | COLOR_PAIR(cfg.curctx + 1));
+#endif
 
 	/* Go to first entry */
 	if (curscroll > 0) {
```

You can use the patch locally if you want.

**Reply by Praful** [reply]
@N-R-K thank you very much for taking the time to provide this. I agree my original suggestion looks noisy. It's not easy to differentiate between entries. However, setting `NNN_COLOR` and removing `<number>=` lets me clearly differentiate between entries: 

![Screenshot from 2024-07-30 16-22-22](https://github.com/user-attachments/assets/4c31e17a-1109-464a-8338-64ffa523c05f)

Obviously, colours can be [changed](https://github.com/jarun/nnn/wiki/Usage#configuration) with `NNN_COLOR`.

The directories have a leading space since I don't know how to keep the active/current formatting and remove the number. Perhaps you can help?!

Could this be provided with a runtime or compile-time option (such as TAB_NAMES=1, default=0) so that I don't have to patch the code every time. Others may find this a useful option.

Thanks again

**Reply by N-R-K** [reply]
> It's not easy to differentiate between entries.

This is only half the problem. The other half is what happens when all these paths overflow the tab width.

![image](https://github.com/user-attachments/assets/6a95852b-6f28-43eb-bbe8-c2d36d12b4b1)

This is just with 2 active tabs and it's already overflowing into the next line. Remember that we offer up to 8 context. Also keep in mind that `nnn` is also used on mobile devices where screen width is small.

> Could this be provided with a runtime or compile-time option (such as TAB_NAMES=1, default=0) so that I don't have to patch the code every time. Others may find this a useful option.

Adding it upstream would require much more change to make sure it works properly without overflowing the line. Shortening paths to basename and/or truncating it beyond a certain width etc. My half-baked patch is no good for upstream inclusion.

**Reply by Praful** [reply]
@N-R-K Fair points. Given this feature will be opt-in, for those who want it, your "half-baked patch" might be good enough as a v1 - perhaps with a bit of tidying up. Of course, a future update could truncate the paths, etc. Personally, this change is a very useful improvement in usability and I'm quite happy living with text occasionally wrapping around.

Good enough now can be better than perfect in the future: features seen as suboptimal by developers can be seen as very useful by users 😊

**Reply by rieje** [reply]
@N-R-K Is it trivial to modify the patch to use basename or e.g. for `~/downloads/pictures/cats` to be `~/d/p/c` or `~/d/p/cats`? that would make it totally sable for me. I've always wanted this feature but knew it wouldn't make sense for nnn to support this for mainline.

**Comment by charbelnicolas** [comment]
I'm launching nnn with the following options through a labwc shortcut on wayland:
```
<keybind key="W-f">
	<action name="Execute" command="foot nnn -xUJHdo" />
</keybind>
```
And the directory is not shown on the title.

Only when launching directly from the terminal will it work. I'm using foot terminal. Would you happen to know why?

**Reply by charbelnicolas** [reply]
All other launch options work but not the -x
